ocaml-lsp-server/src/document.ml

```diff
@@ -142,37 +145,43 @@ let make_config uri =
   Mconfig.get_external_config path mconfig
 
 let make_pipeline thread tdoc =
-  let async_make_pipeline =
-    Scheduler.async_exn thread (fun () ->
-        let text = Text_document.text tdoc in
-        let source = Msource.make text in
-        let config =
-          let uri = Text_document.documentUri tdoc in
-          make_config uri
+  ( lazy (Msource.make (Text_document.text tdoc))
+  , Lazy_fiber.create (fun () ->
+        let async_make_pipeline =
+          Scheduler.async_exn thread (fun () ->
+              let config =
+                let uri = Text_document.documentUri tdoc in
+                make_config uri
+              in
+              Mpipeline.make config (Msource.make (Text_document.text tdoc)))
         in
-        Mpipeline.make config source)
-  in
-  let open Fiber.O in
-  let+ res = await async_make_pipeline in
-  match res with
-  | Ok s -> s
-  | Error e -> Exn_with_backtrace.reraise e
+        let open Fiber.O in
+        let+ res = await async_make_pipeline in
+        match res with
+        | Ok s -> s
+        | Error e -> Exn_with_backtrace.reraise e) )
```

Comment:
seems like we could use `await_exn`

Reply:
await_exn isn't defined in this module and I hate to introduce it. I don't like these little helpers that don't encapsulate any logic.